### ISSUE TITLE
Docker build/push performance is inexplicably slow

### ISSUE THREAD

**Issue by groman2** [docker/docker#28732]
**Description**

I am evaluating whether I could use docker as a reproducible build and runtime environment for a local cluster application. I've got it mostly working, except I've run into the unexpected issue that the build time became excruciatingly slow due to docker image build/push. I've isolated the problem to be related to sheer data volume. 

**Steps to reproduce the issue:**
1. Create test folder
`$ mkdir testimage`
2. Generate 1GB incompressible file
`$ openssl enc -aes128 -in /dev/zero -out >(dd bs=64K count=16384 iflag=fullblock of=testimage/data.img) -e -k $RANDOM -bufsize 134217728`
3. Generate test Dockerfile
`$ echo "FROM scratch" > testimage/Dockerfile`
`$ echo "ADD data.img /data.img" >> testimage/Dockerfile`
4. Build the docker image
`$ /usr/bin/time docker build -t testimage testimage`
5. Tag the image
`$ docker tag testimage localhost:5000/testimage`
6. Push the image to the local registry
`$ /usr/bin/time docker push localhost:5000/testimage`

**Describe the results you received:**
Step (4) output above looks like this:

```
Sending build context to Docker daemon 1.074 GB
Step 1 : FROM scratch
 --->
Step 2 : ADD data.img /data.img
 ---> 9cacb55ce73d
Removing intermediate container 225630bb7d2e
Successfully built 9cacb55ce73d
1.61user 1.85system 0:39.16elapsed 8%CPU (0avgtext+0avgdata 20728maxresident)k
0inputs+0outputs (0major+1658minor)pagefaults 0swaps
```

Which works out to be about **25MB/s**

Step (6) output above looks like this:

```
The push refers to a repository [localhost:5000/testimage]
fa2f5c3d71b4: Pushed
latest: digest: sha256:775f2fdf508b162f5ef26ee96292a9b8c1bbbabe4675e071ce02c1a815cb5845 size: 531
0.31user 0.14system 1:40.09elapsed 0%CPU (0avgtext+0avgdata 20172maxresident)k
0inputs+0outputs (0major+1663minor)pagefaults 0swaps
```
Which works out to be about **10MB/s**

**Describe the results you expected:**

Well it's basically a local copy + hash. There's 64 cores with 2 threads each, 1.5TB of free RAM and nobody else on the system. I expected... a lot faster? Even if it's single-threaded, I would expect maybe >300MB/s for build and >1GB/s for localhost push?

As a point of comparison:
`$ sha1sum data.img`
Takes 2.5 seconds
and
`$ sha256sum data.img`
Takes 5 seconds

Is this because of loop devices?

**Additional information you deem important (e.g. issue happens only occasionally):**

Just for a point of comparison, I decided to try something **really** not designed for handling binaries and do a similar experiment. I copied the data.img to an empty git repo, with a corresponding bare remote on a **different computer** on the local network.

`$ git add data.img`
Took 36s for a pitiful **~27MB/s**
`$ git commit -m "Test"`
Took less than a second
`$ git push srvr4 master`
Took 33.8s for a **~29.5MB/s**

So git add/commit/push over the local network is still faster than docker build/docker push locally, even though docker is a tool nominally designed for blobs of incompressible binary data and shoving those into git is probably the worst thing you can do to it.

**Output of `docker version`:**

```
$ docker --version
Docker version 1.12.3, build 6b644ec
```

**Output of `docker info`:**

```
$ docker info
Containers: 6
 Running: 1
 Paused: 0
 Stopped: 5
Images: 3
Server Version: 1.12.3
Storage Driver: devicemapper
 Pool Name: docker-8:2-3671813-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 11.92 GB
 Data Space Total: 107.4 GB
 Data Space Available: 95.46 GB
 Metadata Space Used: 7.684 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.14 GB
 Thin Pool Minimum Free Space: 10.74 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Use `--storage-opt dm.thinpooldev` to specify a custom block storage device.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.110 (2015-10-30)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: host bridge null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 4.4.0-47-lowlatency
Operating System: Ubuntu 16.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 128
Total Memory: 1.476 TiB
Name: lfvs1
ID: PPFZ:4NG7:WN7F:HTAI:GH3M:UOYE:DFV4:THVD:LRKW:2W4Z:NSPU:TNSP
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Physical box with plenty of ram and CPU, and reasonably fast disks:
```
$ head -n1 /etc/issue
Ubuntu 16.04.1 LTS \n \l
$ uname -a
Linux lfvs1 4.4.0-47-lowlatency #68-Ubuntu SMP PREEMPT Wed Oct 26 21:00:05 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
$ cat /proc/cpuinfo | grep "model name" | cut -d ':' -f 2 | sort | uniq -c
    128  Intel(R) Xeon(R) CPU E7-8867 v3 @ 2.50GHz
$ free -h
              total        used        free      shared  buff/cache   available
Mem:           1.5T        5.5G        1.0T         19M        459G        1.5T
Swap:            0B          0B          0B
```

**Comment by thaJeztah** [comment]
I would definitely not run on devicemapper with loop devices, but not sure that explains everything. w.r.t 64 cores; I wonder if that's actually related to https://github.com/golang/go/issues/15395, but difficult to tell without more metrics

/cc @unclejack @stevvooe

**Comment by groman2** [comment]
Switched docker to using an LVM thinpool on the local RAID
```
# docker info | grep -i "pool\|metadata\|thin"
WARNING: No swap limit support
 Pool Name: docker-thinpool
 Pool Blocksize: 524.3 kB
 Metadata file: 
 Metadata Space Used: 454.7 kB
 Metadata Space Total: 2.986 GB
 Metadata Space Available: 2.986 GB
 Thin Pool Minimum Free Space: 28.37 GB
```
Build performance got worse:
```
 /usr/bin/time docker build -t testimage testimage
Sending build context to Docker daemon 1.074 GB
Step 1 : FROM scratch
 ---> 
Step 2 : ADD data.img /data.img
 ---> d5d415f8f7b7
Removing intermediate container 97b326f08ba7
Successfully built d5d415f8f7b7
1.66user 1.83system 0:48.24elapsed 7%CPU (0avgtext+0avgdata 20460maxresident)k
0inputs+0outputs (0major+1665minor)pagefaults 0swaps
```
A whopping **20.7MB/s**

And local push performance stayed about the same:
```
$ /usr/bin/time docker push localhost:5000/testimage
The push refers to a repository [localhost:5000/testimage]
789e203de5cd: Pushed 
latest: digest: sha256:25909798fce96da752bcdd6bcffc4891c0583264c36ad144b9263c27819274c3 size: 531
0.30user 0.15system 1:41.19elapsed 0%CPU (0avgtext+0avgdata 22184maxresident)k
0inputs+0outputs (0major+1688minor)pagefaults 0swaps
```
~**9.9MB/s**

**Comment by stevvooe** [comment]
@groman2 There are a few things going on here. To understand push performance, you really need to understand what is happening. We'll dive into that then we can look into what is going on with this machine, because it is _very_ slow.

Pushing an image isn't exactly straightforward but we can break down the process into pushing metadata and pushing layers. Since images are dominated by pushing layers, we'll look at what happens with a single layer. Roughly, a layer push consists of the following steps:

1. Tar up the contents of the layer.
2. Compress the resulting tar file.
3. Take the sha256 of the result.
4. Send it up to a registry.

For the most part, this is all done as a "cut-through" operation, meaning we tar directly into the compressor and directly hash the output, sending the hashed data to the registry. This gives the property that the bandwidth of the push is dominated by the _slowest_ component. If everything is working correctly, disks, taring and hashing operate at roughly the same bandwidth, making the push process network limited. The tradeoff is that throughput in one pipeline stage can starve or back off other stages, resulting in slower performance.

For reference, we can measure a few of these steps to get a ballpark estimate of what we should be seeing (none of this is science). On my local machine (macbook pro w/SSD), I got the following as bandwidth numbers for various steps:

| Scenario | Command | Bandwidth |
|---------|------|------|
| Tar up 2GB of files | `tar cf - -C ~/go/src/github.com/docker/docker . | pv > /dev/null` | 917MB/s| | Tar and hash 2GB of files | `tar cf - -C ~/go/src/github.com/docker/docker . | pv | shasum -a256` | 97.9 MB/s |
| Registry push | `time curl -vv -H'Content-Type: application/octet-stream' --data-binary @<(tar cf - -C ~/go/src/github.com/docker/docker .) -XPUT $UPLOAD_URL` | ~300 MB/s |

Note, that rather than using random data, I used real data from a 2GB checkout of docker to make the tar process _even slower_ that what it would be in the example provided here. It includes a generated `data.img`, similar in size the file used in the original examples.

While I agree that push _should_ be faster, expecting >300 MB/s is _very_ optimistic. With unloaded and unlimited network and disk, 100 MB/s is more realistic (and very close to the original numbers I got when do V2 protocol tests), depending on the machine.

Let's also get more realistic push numbers with docker on my local box (docker for mac, with 1.12.3, notoriously slow fs performance). I have a local image of the `docker-dev` build, clocking in ~3GB that took roughly 6:36 to push to a registry running in docker for mac. We get about 8 MB/s. The same push to hub took about the same amount of time. With a different machine, also SSD, however, I get ~18 MB/s, when pushing from local docker to local registry. Clearly, the disk has an effect, but we also have unexplained performance degradation.

Given the above, we can start to really see the pipeline affecting actual bandwidth but we can also see where the bottleneck lies. For the most part, compression and hashing are fixed on an unloaded machine. This leaves the problem at the disk or on the network. Since your push test is going to a local registry instance, we can assume the network is around 125MB/s or more, so it is likely that the disks speed is what is affecting the performance. The test data from my local machine confirms this, but also casts doubts about the theoretical limits of push performance.

__TL; DR__ We can make a few conclusions from the data the analysis above:

1. The disks on @groman2's test machine are _very_ slow. While I estimated _at least_ 200 MB/s using your sha256 example (clearly hash limited), it would be good to get some baseline numbers reading from `/var/lib/docker`, reading directly from the same filesystem as docker. Having this data point will tell us if adjusting your disk setup can help with performance.
2. In theory, we should be getting closer to 100 MB/s of push bandwidth. On the "ideal box" (my second laptop with SSD and native docker), we are only 17 MB/s.

Either way, this warrants more investigation. We'll have to instrument each pipeline stage to see where the actual bottleneck is.

**Comment by groman2** [comment]
I did some speed testing on the machine:

Scenario|Command|Bandwidth
---------|----------|----------
Tar up 2GB of files and transfer over TCP to localhost | `$ nc -l 5001 > /dev/null & tar cf - -C testimage . | pv | nc localhost 5001` | 907MiB/s
RAW TCP Localhost Speed |  `$ iperf -s 2>&1 >/dev/null & iperf -c localhost` | 5.4GiB/s
Read 64GB file | `$ dd if=largefile64GB bs=1M | pv >/dev/null` | 1.32GiB/s
tar up /var/lib/docker | `# tar cf - -C /var/lib/docker . | pv > /dev/null` | 2.03GiB/s
Write big file to /var/lib/docker | `# cat largefile64GB | pv > /var/lib/docker/largefile` | 1.11GiB/s
Tar up 2GB file | `$ tar cf - -C testimage . | pv > /dev/null` | 1.98GiB/s
Tar and sha256 2GB file | `$ tar cf - -C testimage . | pv | shasum -a256` | 165MiB/s


So nothing was really jumping out at me. The tar + sha256 speed is not super impressive. I saw you mentioned compression but your speed tests used uncompressed tar. So I threw in gzip:

Scenario|Command|Bandwidth
---------|----------|----------
Tar + gzip 2GB file | `$ tar cf - -C testimage . | gzip | pv >/dev/null` | 31.9MiB/s

I think we found our culprit. `gzip` is slow.... sloooow. Using `gzip --fast` helps only marginally.

Trying the same experiment with `pigz` instead of `gzip` produces some promising results:

Scenario|Command|Bandwidth
---------|----------|----------
Tar + pigz 2GB file | `$ tar cf - -C testimage . | pigz | pv >/dev/null` | 369MiB/s


So it would be nice to have:

- Faster/parallel compression algorithm
- A no-compress option
- A faster hash algorithm 

I guess the sha would have to move prior to compression, because otherwise --no-compress would change the sha

**Comment by stevvooe** [comment]
@groman2 Perfect! This data helps out a lot with the analysis. I think we can eliminate the hypothesis that your "disks are slow".

Throwing out gzip in the analysis above was a __HUGE__ omission. Apologies for the oversight. I've confirmed the bandwidth numbers of gzip, getting anywhere from 10-40 MB/s, even with pigz, on my local machine (I only have a few cores, so this makes sense).

For docker, parallel gzip presents some serious performance issues. Most benchmarks assume that the process will have the whole machine at their disposal. However, with applications running and other layers being push/pulled concurrently, many of these parallel gains will be diminished. We really need to optimize for single threaded performance and gzip seems to be a massive bottleneck.

Looking back, gzip compression has always been a controversial topic. Checkout https://github.com/docker/docker/issues/10181 and https://github.com/docker/docker/pull/17674. The approach previously has always been about optimizing gzip, usually at the cost of more resources. Unfortunately, none of these analysis approached the problem from a theoretical pipeline perspective.

The main reason we cannot just turn gzip on and off is that the compression is part of the content addressable artifact. If you disable compression, it will change the content addresses of existing content.

We'll have to make a decision to as a whole and have a plan for phasing out compressed layers. This will get rid of a major bottleneck and allow operators to choose a fitting transport compression that works well in their infrastructure.

Another option is to compress each layer _once_, rather than on every single push and pull.

**Comment by thaJeztah** [comment]
w.r.t. making compression optional, there's also a discussion here; https://github.com/docker/docker/issues/1266

**Comment by groman2** [comment]
I'm not familiar with the Docker codebase, but I have some familiarity with content addressable systems in general. In my experience, generating a hash of compressed data for content addressability can result in an eventual mess that is not related to performance. You are effectively locking yourself to a particular compression library version/compression settings/etc.

while we have a guarantee that
```
$ cat data.img | gzip | gunzip | shasum
```
and
```
$ shasum data.img
```
will give the same result on any system, we don't have a guarantee that the compressed version will have the same SHA everywhere and every time (as long as it decompresses into the same thing, it's all fine). Obviously it's likely that it IS the same, given that gzip is pretty stable.

Using the sha of the _uncompressed_ data for content addressability is a much better long term approach, IMO, even if compressed layers are supported.

**Comment by stevvooe** [comment]
@groman2 I see you are very familiar with this trade off. The actual goal with this design was to make the storage layers agnostic to what content is being stored. When we introduced content addressability, the decision had already been made that layers are compressed artifacts, which was affirmed in the presence of possible, though improbable attacks, involving the processing of unverified content.

The root issue here is the expectation that a round trip of content through serialization layers provides a stable hash calculation. In your example, only `data.img` is considered. However, that may include json generation, consistent tar creation (see tar-split) and consistent compression. Whether you have choice of compression or not isn't particularly important when the round tripping of artifacts can introduce hash instability. In particular, bugs within any of these components can create problems with the round trip. Generating the artifact _once_ usually assuages these problems and is very much a viable solution in this case.

Put in your terms, I expand the warning of "generating a hash of compressed data for content addressability can result in an eventual mess" to the stronger "generating hash stable round trip serialization for content addressability _will_ result in an eventual mess". 😉 

The implication here is that compression is _actually_ a transport concern. Especially since we do it on _every_ push and pull. Going back to our bandwidth pipeline analysis, if the compression bandwidth is higher than the connection bandwidth, we have the property that the transfer time (there is a relationship that can be worked out mathematically). For most internet connections and CPU sizes, gzip is adequate because it can keep the pipe full. When the CPU size (lower compression bandwidth) is too small or the connection is too fast (like in this case), the low compression bandwidth starves the network connection, resulting in a slow push. Because the performance of compression is variant on environmental conditions, it should probably be separated from content identity.

I hope my statements above give you the feeling not that I disagree with you, but rather that I am affirming your intuition with experience we've had in the deployment of content addressability in Docker.

Given the above, there are too paths:

1. Separate compression from content identity. We still have the "danger" of hash instability, since we are round tripping the other components, but we have much better control over their stability. This comes with the expense of a migration, but the cost isn't too bad, since daemons have always had content compression detection in layer processing.
2. Don't round trip artifacts. This ensures stability for all serialization methods in the generation of an image. At the cost of disk space and a compression step during commit, it will store content, as received. If content is to be pushed again, the untouched data will be sent directly without interference from the docker daemon.

Both can be done, but we can favor 2 before 1, since there will be less impact. The takeaway is that I am going to pursue these much less casually than before, since we actually have some theory backing it up.

**Comment by D0han** [comment]
Is this being worked on? Pushing 16GB of layers on beefy server takes very, very long time.

**Comment by Tomasu** [comment]
Very interested in seeing this improved. It takes a long time to push to local registries on a lan that can easily handle full GiB throughput.

**Comment by stevvooe** [comment]
@D0han @Tomasu I am generally addressing this problem with the approach taken in containerd but I am not sure how quickly we can realize those gains in docker.

The quickest fix would be to allow one to completely disable layer compression.

**Comment by D0han** [comment]
Thanks @stevvooe !
Is disabling doable by options, or not implemented yet?

**Comment by stevvooe** [comment]
> Is disabling doable by options, or not implemented yet?

It is not implemented.

**Comment by Tomasu** [comment]
@stevvooe I see. I really don't mind that too much. disk space is cheap on the registry for what I use it for.

**Comment by stevvooe** [comment]
@Tomasu I am currently working on the option to enable saving of artifacts. Give [containerd](https://github.com/containerd/containerd/blob/master/docs/getting-started.md) a try on a pull-push round trip and see if that impacts the push time and we can adjust. It is not quite production ready yet, but it will be a good data point to ensure the approach is right.

**Comment by farizluqman** [comment]
any updates?

**Comment by berlinguyinca** [comment]
any updates, been 2 years and the issue still exists.

**Comment by SkyaTura** [comment]
any updates?

**Comment by shreyanshc** [comment]
@thaJeztah @stevvooe 

Is there a way to calculate total time spent compressing / decompressing images during push/pull operations (by inserting logs in moby source) , rather than relying on "time docker push" as this includes network time also ? 

I am trying to analyze how much more time does the gzip/unpigz spend on comp/decomp when used from within docker , and when we compress the same thing using docker save <image> | gzip | ... (basically want to measure overhead due to stream compress/decompress operations in docker )

**Comment by thaJeztah** [comment]
If you have debug enabled, the daemon will show the logs for each step;

```
DEBU[2019-08-07T19:30:12.680669875Z] Calling HEAD /_ping                          
DEBU[2019-08-07T19:30:12.694678713Z] Calling GET /v1.40/info                      
DEBU[2019-08-07T19:30:12.810109275Z] Calling POST /v1.40/images/create?fromImage=ubuntu&tag=latest 
DEBU[2019-08-07T19:30:12.817316208Z] Trying to pull ubuntu from https://registry-1.docker.io v2 
DEBU[2019-08-07T19:30:14.662277902Z] Pulling ref from V2 registry: ubuntu:latest  
DEBU[2019-08-07T19:30:14.662347404Z] docker.io/library/ubuntu:latest resolved to a manifestList object with 6 entries; looking for a unknown/amd64 match 
DEBU[2019-08-07T19:30:14.662364930Z] found match for linux/amd64 with media type application/vnd.docker.distribution.manifest.v2+json, digest sha256:d91842ef309155b85a9e5c59566719308fab816b40d376809c39cf1cf4de3c6a 
DEBU[2019-08-07T19:30:15.134344421Z] pulling blob "sha256:1d425c98234572d4221a1ac173162c4279f9fdde4726ec22ad3c399f59bb7503" 
DEBU[2019-08-07T19:30:15.134372035Z] pulling blob "sha256:7413c47ba209e555018c4be91101d017737f24b0c9d1f65339b97a4da98acb2a" 
DEBU[2019-08-07T19:30:15.134405110Z] pulling blob "sha256:0fe7e7cbb2e88617d969efeeb3bd3125f7d309335c736a0525233ec2dc06aee1" 
DEBU[2019-08-07T19:30:15.656331038Z] Downloaded 1d425c982345 to tempfile /var/lib/docker/tmp/GetImageBlob813512277 
DEBU[2019-08-07T19:30:15.656536674Z] pulling blob "sha256:344da5c95cecd0f55238ce59b8469ee301056001ece2b769e9691b80f94f9f37" 
DEBU[2019-08-07T19:30:15.680269873Z] Downloaded 0fe7e7cbb2e8 to tempfile /var/lib/docker/tmp/GetImageBlob175359311 
DEBU[2019-08-07T19:30:16.285710722Z] Downloaded 344da5c95cec to tempfile /var/lib/docker/tmp/GetImageBlob741678178 
DEBU[2019-08-07T19:30:19.963628678Z] Downloaded 7413c47ba209 to tempfile /var/lib/docker/tmp/GetImageBlob377162672 
DEBU[2019-08-07T19:30:19.968940733Z] Applying tar in /var/lib/docker/overlay2/e627c5b69cb82a1d0f77357522926a0ef1351be2b20d17d3937b4ad9a6e901ad/diff  storage-driver=overlay2
DEBU[2019-08-07T19:30:21.400568017Z] Applied tar sha256:543791078bdb84740cb5457abbea10d96dac3dea8c07d6dc173f734c20c144fe to e627c5b69cb82a1d0f77357522926a0ef1351be2b20d17d3937b4ad9a6e901ad, size: 63197287 
DEBU[2019-08-07T19:30:21.461786503Z] Applying tar in /var/lib/docker/overlay2/dd94d8f18f96e8264f9898cd63e6037e880f1ebf32baff212e11350bdba8d4a2/diff  storage-driver=overlay2
DEBU[2019-08-07T19:30:21.586150772Z] Applied tar sha256:c56e09e1bd18e5e41afb1fd16f5a211f533277bdae6d5d8ae96a248214d66baf to dd94d8f18f96e8264f9898cd63e6037e880f1ebf32baff212e11350bdba8d4a2, size: 987475 
DEBU[2019-08-07T19:30:21.601487548Z] Applying tar in /var/lib/docker/overlay2/a7123da831bb70cab3492f7f44eb31086becde37500001d4b03553d14e354638/diff  storage-driver=overlay2
DEBU[2019-08-07T19:30:21.737944764Z] Applied tar sha256:a31dbd3063d77def5b2562dc8e14ed3f578f1f90a89670ae620fd62ae7cd6ee7 to a7123da831bb70cab3492f7f44eb31086becde37500001d4b03553d14e354638, size: 745 
DEBU[2019-08-07T19:30:21.755832870Z] Applying tar in /var/lib/docker/overlay2/c82433d5d0c2fce81e72606e5f2d8a5bea72fcd7667aa7220e09211d6fcc8e76/diff  storage-driver=overlay2
DEBU[2019-08-07T19:30:21.908411096Z] Applied tar sha256:b079b3fa8d1b4b30a71a6e81763ed3da1327abaf0680ed3ed9f00ad1d5de5e7c to c82433d5d0c2fce81e72606e5f2d8a5bea72fcd7667aa7220e09211d6fcc8e76, size: 7 
```

**Comment by shreyanshc** [comment]
Thank you for the reply. 

I am currently using debug mode, and it has timestamps for "Applying tar" and "Applied tar" for pull (decompression).  But I was unsure if this is accurate, as the debug logging code is spread across 2 files :
[daemon/graphdriver/overlay2/overlay.go ---- line 712](https://github.com/moby/moby/blob/ed07e1152879a4d156dff2e86abca3c4c811e743/daemon/graphdriver/overlay2/overlay.go#L712) logger.Debugf("Applying tar in %s", applyDir) 

[layer/layer_store.go line 277](https://github.com/moby/moby/blob/66f8f2b87c7bca5e1f99bd61eaa5165524bee14d/layer/layer_store.go#L277)----- logrus.Debugf("Applied tar %s to %s, size: %d", layer.diffID, layer.cacheID, applySize)





But there is no end point specified for push-compression finish of a layer.

You have "Assembling tar data" & then a "uploaded layer " , which has a timestamp for after upload finishes, rather than after compression finishes.

I tried to add debug logs before and after this [switch](https://github.com/moby/moby/blob/4866f5139a0b0bc026aa4b0c916ee1071c695134/distribution/push_v2.go#L452) for compression times in push_v2.go, but the logs show almost same timestamps.

I am trying to find between which 2 points does compression and decompression part happen in moby code. Appreciate your time and help, thank you

**Comment by thaJeztah** [comment]
I'm not super-familiar with that part of the code, but happy to help where I can

**Comment by shreyanshc** [comment]
Will it be possible to point me to someone who is familiar with it?

**Comment by thaJeztah** [comment]
@dmcgowan @stevvooe and @tonistiigi have dealt more with that code

**Comment by Alveona** [comment]
still having the issue from time to time 🥴

**Comment by nicolamori** [comment]
Also here, a 3 min (almost fully cached) build that takes up to 30 min for being pushed. It's weird.

**Comment by da2018** [comment]
> I would definitely not run on devicemapper with loop devices, but not sure that explains everything. w.r.t 64 cores; I wonder if that's actually related to [golang/go#15395](https://github.com/golang/go/issues/15395), but difficult to tell without more metrics
> 
> /cc @unclejack @stevvooe

Hi, could you explain why running on devicemapper with loop devices is a bad idea? Thanks.

**Comment by thaJeztah** [comment]
@da2018 various reasons can be found in the docs here; https://docs.docker.com/storage/storagedriver/device-mapper-driver/#device-mapper-and-docker-performance

Unless you have a very specific reason, I'd recommend using the default (`overlay2`) storage driver. The device mapper storage driver was created when RHEL/CentOS kernels did not yet support overlayfs or other alternatives, but now that they do, overlay2 is used on those platforms as well. For those reasons, the device mapper storage driver [is marked deprecated](https://docs.docker.com/engine/deprecated/#device-mapper-storage-driver); there's no immediate plans to _remove_ it, but other than some maintenance if needed, no further development is planned.

**Comment by da2018** [comment]
> ry specific reason, I'd recommend using the default (`overlay2`) storage driver. The device mapper storage driver was created when RHEL/CentOS kernels did not yet support overlayfs or other alternatives, but now that they do, overlay2

Many thanks.

**Comment by Richard004** [comment]
Deleting LOCAL docker images helped in my case!!

Maybe unrelated, but my cca. 300MB push suffered the same terrible slowdown. Used to be seconds, but became long minutes (after few weeks of work).
So I tried a few things (like disabling antivirus for Docker), but what seems to have helped was removing all images from my local docker. It freed up some 7GB and the push is suddenty super fast again.

**Comment by umialpha** [comment]
buildah pushing is much faster than docker-cli because it uses "github.com/klauspost/pgzip" instead of "compress/gzip".

According to the pgzip benchmark, pgzip is 123x faster than gzip.

Is there anyone interested to send a pr for mody?

**Comment by pablodz** [comment]
Still a problem, very frequently

**Comment by thetredev** [comment]
Still an issue with Docker 24.0.5. I second @umialpha.

**Comment by Pazekal90** [comment]
We have the same problems.

**Comment by Benjamin-Tan** [comment]
We have the same problems. Is there any way to speed this up?

**Comment by thetredev** [comment]
@Benjamin-Tan yes, like @umialpha said: using `github.com/klauspost/pgzip` instead of `compress/gzip`. I guess the Moby team hasn't heard us yet.

**Comment by Benjamin-Tan** [comment]
@thetredev How would one go about to use pgzip instead of gzip? while waiting for the Moby team to recognize this.

**Comment by thetredev** [comment]
@Benjamin-Tan I would imagine: figuring out how to build moby, the Docker engine, and the CLI. Not sure if the Docker engine is open source (it's free though), so you might be able to replace the Moby binaries from the engine with those you built yourself.

I guess I'm not the right person to ask though. It could very likely be the case that you have to build containerd instead of Moby, with the gzip patch applied to either Moby or containerd (don't know how they work together, if at all). But if it's only containerd then it's easier because at least on Linux, the Docker CLI uses containerd AFAIK. Again, not sure what role Moby plays in all of this. It might be best to read through the containerd project first and go from there.

**Comment by thetredev** [comment]
@benatkin my 2 cents: the actual performance gain depends on how the resulting image is structured before compression. Layering also seems to be very important. There's only so much parallelism can do. Figuring out the perfect layering for speeding up compression is up to you. So 123% performance increase is mist likely achievable, but it could go down quickly if image layering is not in favor of compression.

**Comment by thaJeztah** [comment]
We're working on completing the containerd image-store integration, which switches the image store to use the image-store and snapshotters provided by containerd.

With the containerd store integration, compressed (tar.gz) layers are persisted on disk (instead of constructed when pushing); this allows pushes to be reproducible (compression is not 100% reproducible), but at the cost of additional storage space required. If you have a system to test on, you can enable it in current versions of docker; see https://docs.docker.com/storage/containerd/#containerd-image-store-with-docker-engine

**Comment by Benjamin-Tan** [comment]
@thaJeztah thank you! I had a try and it is a lot faster. I did not even notice that it has been pushed successfully. When do you think this feature would become General Availability, away from Beta stage?

The solution might benefit #26959

**Comment by thaJeztah** [comment]
Unless I'm mistaken, I don't think docker engine is used at all when using the buildx Kubernetes driver; in that case, you're running a standalone BuildKit builder (which doesn't store images, only build-cache).

**Comment by thetredev** [comment]
Ahh thank makes sense. Thank you for clearing it up.

**Comment by Benjamin-Tan** [comment]
@thaJeztah for your information, with this experimental feature enabled, the user permission of folders does not persist in the running container. Also, the docker image size increase drastically (from 2.78 GB to 9 GB) just for a few line in the dockerfile to add new user,group via useradd. Not sure where would be a good place to file this, but feel free to move this information elsewhere if that helps!

A simple Dockerfile can be found as follows:
```
FROM ubuntu:jammy as base

RUN groupadd -g 1000 docker_group &&\
    useradd -l -m -d /home/docker_user -u 1000 -g docker_group docker_user -s /bin/bash -G sudo -p "" &&\
    install -d -m 0755 -o docker_user -g docker_group /home/docker_user

RUN install -o docker_user -g docker_group -d /folderD/ && \
    install -o docker_user -g docker_group -d /home/docker_user/folderA && \
    install -o docker_user -g docker_group -d /home/docker_user/.folderB

RUN chown -R docker_user:docker_group /home/docker_user/
RUN ls -la /home/docker_user
USER docker_user:docker_group

WORKDIR /home/docker_user/folderA
```

> We're working on completing the containerd image-store integration, which switches the image store to use the image-store and snapshotters provided by containerd.
> 
> With the containerd store integration, compressed (tar.gz) layers are persisted on disk (instead of constructed when pushing); this allows pushes to be reproducible (compression is not 100% reproducible), but at the cost of additional storage space required. If you have a system to test on, you can enable it in current versions of docker; see https://docs.docker.com/storage/containerd/#containerd-image-store-with-docker-engine

**Comment by rumpl** [comment]
@Benjamin-Tan I just built an image with the Dockerfile you gave and:

```console
REPOSITORY                                                                   TAG       IMAGE ID       CREATED             SIZE
huge                                                                         latest    50f780ba3a48   5 seconds ago       117MB
```

How did you get a 9GB image?

If you have a reproducer feel free to open a new issue

**Comment by Benjamin-Tan** [comment]
> @Benjamin-Tan I just built an image with the Dockerfile you gave and:
> 
> ```
> REPOSITORY                                                                   TAG       IMAGE ID       CREATED             SIZE
> huge                                                                         latest    50f780ba3a48   5 seconds ago       117MB
> ```
> 
> How did you get a 9GB image?
> 
> Also, I see that the permissions are right it would seem
> 
> ```
> docker_user@a6a81fba0b78:~$ pwd
> /home/docker_user
> docker_user@a6a81fba0b78:~$ ls -la
> total 28
> drwxr-xr-x 1 docker_user docker_group 4096 Jan  4 10:31 .
> drwxr-xr-x 1 root        root         4096 Jan  4 10:31 ..
> -rw-r--r-- 1 docker_user docker_group  220 Jan  6  2022 .bash_logout
> -rw-r--r-- 1 docker_user docker_group 3771 Jan  6  2022 .bashrc
> drwxr-xr-x 2 docker_user docker_group 4096 Jan  4 10:31 .folderB
> -rw-r--r-- 1 docker_user docker_group  807 Jan  6  2022 .profile
> drwxr-xr-x 2 docker_user docker_group 4096 Jan  4 10:31 folderA
> ```
> 
> If you have a reproducer feel free to open a new issue

I was extracting this out of another Dockerfile with multi-stage build. I will get a reproducible example and open the new issue instead. Thanks @rumpl

**Comment by matti** [comment]
yes, it would be solved here https://github.com/moby/moby/issues/28732#issuecomment-1842582615 if anybody would care

**Comment by Paebbels** [comment]
It's disappointing to see this problem still exists and isn't addressed by Docker. Meanwhile software and Docker images have grown. I need to deal with a 60..100 GB image containing a "compiler". Unfortunately  not everything is as small as a 350 MB GCC...

Rebuilding all supported SDK versions takes me currently a 3 days! All is based on a 20G network, SSDs everywhere and latest >64 core CPUs. And then there is super slow Docker push...

From reading the comments, I don't get, why the algorithm can't be fixed or replaced or made switchable. 8 years passed by, with plenty of time to have e.g. multiple algorithms (e.g. for the hash) in parallel, so the other could be phased out. But it looks like no action was taken and we're still at the beginning of the discussion.

I'll now investigate how to slice a Docker image into equally sized 1G layers, so maybe it can compute and push in parallel. I already measure a significant boost for Docker pull, because someone forgot to implement parallel downloads via HTTP in Docker pull.

**Comment by lahwaacz** [comment]
@Paebbels In 2025 at least we have functional alternatives like buildah and podman

**Comment by Paebbels** [comment]
I found this hidden in the documentation: https://docs.docker.com/build/exporters/#configuration-options

While it lists, "zstd", it's not the "real" zstd implementation, but a reimplementation in Go with less speedup and no multi-threading support...

Moreover, it's only documented for `build` but not for `push`.  
So it doesn't cover all use cases like:
1. building an image
2. running that image to get the installed software version of a package (not known ahead)
3. retagging that temp build to the actual tag
4. pushing that tag to a on-premise registry

It also demolishes that docker offers multiple commands as building pieces for a bigger goal (like RISC), but now we are forced to use a single `CISC` like command to use such a feature.

Does anyone know if image registries like GitLab Container Registry accept such zstd compressed layers?

**Comment by thetredev** [comment]
@Paebbels seems so, at least they say they do: https://gitlab.com/gitlab-org/container-registry/-/blob/master/docs/supported-media-types.md

**Comment by Paebbels** [comment]
Thanks @thetredev .

-----
Today I did measurements with gzip and zstd compression. I can tell for sure, there is no `zstd`, because two different compression algorithms compressing 54 GB of data with >100k files can NOT result in two equalized images !

I'll now test the compression level parameter.

**Comment by thetredev** [comment]
@Paebbels yeah I've played around with compression levels using GitLab 15 I think it was. Didn't go very far.. maybe they've fixed it until now.

Looking forward to see your results!

**Comment by Paebbels** [comment]
@thetredev 
Like selecting the compression algorithm, the compression level has no effect at all. Thus I conclude, there is just a single algorithm used by Docker. The Docker documentation is not correct, I would almost conclude it's all fake about compression. Maybe it was planned, but for sure: it's not working at all.

It's also silly both from Go and from Klaus Post to reimplement gzip and zstd purely in Go and completely mess up the feature set as well as the performance. All target platforms provide libz and libzstd for integration. Why reinvent the wheel and cripple the performance?

-----
Besides bad performance in compression/decompression, Docker faces more mischiefs like sequential uploads and downloads at push and pull operations (Docker daemon needs to be reconfigured to >10 parallel connection to achieve 1Gb/s, not talking yet about my 10G CI network) and it uses no parallelism for compression/decompression of individual overlays, which could run all in parallel.

I think I should summarize all findings in a bigger article. It's frightening to see the world relying on Docker with such a stupid implementation. It's a waste of network and CPU resources because everything runs sequential and "single threaded" ...

It would be interesting to know if K8s uses these functions too, or if they have there own build, push and pull algorithms as well as compression/decompression. If not, we could speed up all clouds and especially CI setups with Docker images exchanged every minute ...

**Comment by Paebbels** [comment]
Addendum, were should I open an issue for my findings?  

At moby, at Docker CLI, Buildx, buildkit, ...
I can't find a dockerd repository.

**Comment by thompson-shaun** [comment]
👋 @Paebbels 

The [moby/moby](https://github.com/moby/moby/issues) project would be the place to start.

**Comment by slonopotamus** [comment]
We're already there.
